docs/dev/stream/state/broadcast_state.md

```diff
@@ -0,0 +1,279 @@
+    // we keep a list as we may have many first elements waiting
+    private final MapStateDescriptor<String, List<Item>> mapStateDesc =
+	    new MapStateDescriptor<>(
+	        "items",
+	        BasicTypeInfo.STRING_TYPE_INFO, 
+	        new ListTypeInfo<>(Item.class));
+
+    // identical to our ruleStateDescriptor above
+    private final MapStateDescriptor<String, Rule> ruleStateDescriptor = 
+        new MapStateDescriptor<>(
+    	    "RulesBroadcastState",
+    		BasicTypeInfo.STRING_TYPE_INFO,
+    		TypeInformation.of(new TypeHint<Rule>() {}));
+
+	@Override
+	public void processBroadcastElement(Rule value, 
+	                                    Context ctx, 
+	                                    Collector<String> out) throws Exception {
+	    ctx.getBroadcastState(ruleStateDescriptor).put(value.name, value);
+	}
+
+	@Override
+	public void processElement(Item value, 
+	                           ReadOnlyContext ctx, 
+	                           Collector<String> out) throws Exception {
+
+        final MapState<String, List<Item>> state = getRuntimeContext().getMapState(mapStateDesc);
+        final Shape shape = value.getShape();
+    
+        for (Map.Entry<String, Rule> entry: 
+                ctx.getBroadcastState(ruleStateDescriptor).immutableEntries()) {
+            final String ruleName = entry.getKey();
+            final Rule rule = entry.getValue();
+    
+            List<Item> stored = state.get(ruleName);
+            if (stored == null) {
+                stored = new ArrayList<>();
+            }
+    
+            if (shape == rule.second && !stored.isEmpty()) {
+                for (Item i : stored) {
+                    out.collect("MATCH: " + i + " - " + value);
+                }
+                stored.clear();
+            }
+    
+            // there is  no else{} to cover if rule.first == rule.second
+            if (shape.equals(rule.first)) {
+                stored.add(value);
+            }
+    
+            if (stored.isEmpty()) {
+                state.remove(ruleName);
+            } else {
+                state.put(ruleName, stored);
+            }
+        }
+	}
+}
+{% endhighlight %}
+
+## Important Considerations
+
+After describing the offered APIs, this section focuses on the important things to keep in mind when using broadcast 
+state. These are:
+
+  - **There is no cross-task communication:** As stated earlier, this is the reason why only the broadcast side of a 
+`(Keyed)-BroadcastProcessFunction` can modify the contents of the broadcast state. In addition, the user has to make 
+sure that all tasks modify the contents of the broadcast state in the same way for each incoming element. Otherwise,
+different tasks might have different contents, leading to inconsistent results.
+
+  - **Order of events in Broadcast State may differ across tasks:** Although broadcasting the elements of a stream 
```

Comment:
Maybe we should simply clearly state that, the state update for each incoming element **CANNOT** be order-depending.